src/Server/TestKeeperTCPHandler.cpp

```diff
@@ -332,41 +325,31 @@ void TestKeeperTCPHandler::runImpl()
             if (close_received)
                 break;
 
-            if (state & SocketInterruptablePollWrapper::HAS_RESPONSE)
+            for (size_t response_id : result.ready_responses)
             {
-                while (!responses.empty())
-                {
-                    if (responses.front().wait_for(0s) != std::future_status::ready)
-                        break;
+                auto response_future = responses.find(response_id);
+                if (response_future == responses.end())
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Trying to get unknown response #{}", response_id);
 
-                    auto response = responses.front().get();
+                if (response_future->second.wait_for(0s) != std::future_status::ready)
+                    throw Exception(ErrorCodes::LOGICAL_ERROR, "Response #{} was market as ready but corresponding future not ready yet", response_id);
+
+                auto response = response_future->second.get();
+                if (response->error == Coordination::Error::ZOK)
+                {
                     response->write(*out);
-                    responses.pop();
                 }
-            }
-
-            if (state & SocketInterruptablePollWrapper::HAS_WATCH_RESPONSE)
-            {
-                for (auto it = watch_responses.begin(); it != watch_responses.end();)
```

Comment:
The race condition may happen here. We just iterating over all watches and the corresponding request can be already processed but not sent to the client. In this case, we will send a watch response before request response.

Reply:
Actually, this code quite strange and was written as a prototype... after that, I've dug into epoll, compatibility and so on and forgot about it.